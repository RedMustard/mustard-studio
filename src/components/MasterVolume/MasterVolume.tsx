import { h } from 'preact';
import { useContext, useEffect } from 'preact/hooks';
import { setMasterGainNode, handleChangeMasterVolume } from '../../lib/studioService/studioServiceActions';
import { StudioServiceContext } from '../../lib/studioService/StudioServiceStore';
import { VolumeFader } from '../VolumeFader/VolumeFader';

interface MasterVolumeProps {
    audioContext: AudioContext;
}

export const MasterVolume = ({ audioContext }: MasterVolumeProps) => {
    const [studioService, dispatch] = useContext(StudioServiceContext);
    const masterVolume = studioService.volume.master;
    let masterGainNode = studioService.gainNodes.master;

    useEffect(() => {
        if (!masterGainNode) {
            masterGainNode = audioContext.createGain();
            masterGainNode.connect(audioContext.destination);
            setMasterGainNode(masterGainNode, dispatch);
        } else {
            masterGainNode.gain.value = masterVolume;
        }
    }, [masterGainNode, masterVolume]);

    return (
        <div className="master-volume">
            M: &nbsp;
            <VolumeFader
                classSuffix="master"
                value={masterVolume}
                onChange={(value) => handleChangeMasterVolume(value, dispatch)}
            />
        </div>
    );
};
