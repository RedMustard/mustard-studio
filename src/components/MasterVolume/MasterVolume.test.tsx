import { h } from 'preact';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { MasterVolume } from './MasterVolume';

const wamock = require('web-audio-mock-api');

const audioContext = new wamock.AudioContext();

describe('<MasterVolume />', () => {
    const baseProps = {
        audioContext,
    };

    it('renders with basic props', () => {
        const wrapper = shallow(
            <MasterVolume
                {...baseProps}
            />,
        );

        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
