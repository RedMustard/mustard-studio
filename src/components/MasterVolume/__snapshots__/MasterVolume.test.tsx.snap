// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MasterVolume /> renders with basic props 1`] = `
<div
  className="master-volume"
>
  M Vol: Â 
  <VolumeFader
    classSuffix="master-volume"
    onInput={[Function]}
    onMouseDown={[Function]}
    value={0.1}
  />
</div>
`;
